tests/backup_restore_test.go

```diff
@@ -0,0 +1,262 @@
+// 5. lock tables and atomically swap original and ghost tables
+func TestPITR(t *testing.T) {
+	t.Parallel()
+	a := require.New(t)
+
+	localhost := "127.0.0.1"
+	port := getTestPort(t.Name())
+	username := "root"
+	database := "backup_restore"
+
+	insertData := func(db *sql.DB, a *require.Assertions, begin, end int) {
+		tx, err := db.Begin()
+		a.NoError(err)
+		defer tx.Rollback()
+
+		for i := begin; i < end; i++ {
+			_, err := tx.Exec(fmt.Sprintf("INSERT INTO t00 VALUES (%d)", i))
+			a.NoError(err)
+			_, err = tx.Exec(fmt.Sprintf("INSERT INTO t10 VALUES (%d, %d)", i, i))
+			a.NoError(err)
+			_, err = tx.Exec(fmt.Sprintf("INSERT INTO t11 VALUES (%d, %d)", i, i))
+			a.NoError(err)
+		}
+
+		err = tx.Commit()
+		a.NoError(err)
+	}
+
+	_, stop := mysql.SetupTestInstance(t, port)
+	defer stop()
+
+	db, err := sql.Open("mysql", fmt.Sprintf("root@tcp(localhost:%d)/mysql", port))
+	a.NoError(err)
+	defer db.Close()
+
+	_, err = db.Exec(fmt.Sprintf("CREATE DATABASE %s", database))
+	a.NoError(err)
+
+	_, err = db.Exec(fmt.Sprintf("USE %s", database))
+	a.NoError(err)
+
+	_, err = db.Exec(`
+	CREATE Table t00 (
+		id INT,
+		PRIMARY KEY (id),
+		CHECK (id > -1)
+	);
+	`)
+	a.NoError(err)
+	_, err = db.Exec(`
+	CREATE Table t10 (
+		id INT,
+		pid INT,
+		PRIMARY KEY (id),
+		UNIQUE INDEX (pid),
+		CONSTRAINT FOREIGN KEY (pid) REFERENCES t00(id) ON DELETE NO ACTION
+	);
+	`)
+	a.NoError(err)
+	_, err = db.Exec(`
+	CREATE Table t11 (
+		id INT,
+		pid INT,
+		PRIMARY KEY (id),
+		UNIQUE INDEX (pid),
+		CONSTRAINT FOREIGN KEY (pid) REFERENCES t00(id) ON DELETE NO ACTION
+	);
+	`)
+	a.NoError(err)
+
+	// insert data to make time point t0
+	insertData(db, a, 0, 10)
+
+	// make a full backup of t0
+	logger, err := zap.NewDevelopment()
+	a.NoError(err)
+	driver, err := dbPlugin.Open(
+		context.TODO(),
+		dbPlugin.MySQL,
+		dbPlugin.DriverConfig{Logger: logger},
+		dbPlugin.ConnectionConfig{
+			Host:      localhost,
+			Port:      fmt.Sprintf("%d", port),
+			Username:  username,
+			Password:  "",
+			Database:  database,
+			TLSConfig: dbPlugin.TLSConfig{},
+		},
+		dbPlugin.ConnectionContext{},
+	)
+	defer driver.Close(context.TODO())
+
+	var buf bytes.Buffer
+	err = driver.Dump(context.TODO(), database, &buf, false)
+	a.NoError(err)
+	t.Log(buf.String())
+
+	// insert more data to make time point t1
+	insertData(db, a, 100, 200)
+
+	// concurrently insert data
+	stopChan := make(chan bool)
+	go func() {
+		i := 200
+
+		db, err := sql.Open("mysql", fmt.Sprintf("root@tcp(localhost:%d)/mysql", port))
+		a.NoError(err)
+		defer db.Close()
+
+		_, err = db.Exec(fmt.Sprintf("USE %s", database))
+		a.NoError(err)
+
+		t.Log("Start inserting data")
+		for {
+			select {
+			case <-stopChan:
+				t.Log("Stop inserting data")
+				return
+			default:
+			}
+			insertData(db, a, i, i+1)
+			time.Sleep(10 * time.Millisecond)
+		}
+	}()
+	defer func() { stopChan <- true }()
+
+	// restore to ghost tables
+	_, err = db.Exec("SET foreign_key_checks=OFF")
```

Comment:
There are a couple of common pieces we may extract to be shared across different methods.

1. Make some changes. There are 2 types of changes. DDL and DML. And for DML, I am thinking of constructing a single row and use UPDATE statement to keep updating that statement. Thus it's easy to return the current state and used for data validation later.
2. Do a backup
3. Do a restore
4. Validate data.

Methods like "db.Exec("SET foreign_key_checks=OFF")" are specific to the test on the other hand.